pelican/plugins/sitemap/sitemap.py

```diff
@@ -55,235 +58,104 @@ def format_date(date):
-                )
-                standard_page_save_as = self.context.get(
-                    "{}_SAVE_AS".format(standard_page.upper())
-                )
-                fake = FakePage(
-                    status="published",
-                    date=self.now,
-                    url=standard_page_url or "{}.html".format(standard_page),
-                    save_as=standard_page_save_as or "{}.html".format(standard_page),
+            assert changefreq in VALID_CHANGEFREQS
+            priority = float(
+                context.get("SITEMAP", {})
+                .get("priorities", {})
+                .get(content_type, PRIORITY_DEFAULTS[content_type])
+            )
+            file.write(
+                XML_URL_TEMPLATE.format(
+                    loc=abs_url,
+                    lastmod=format_date(lastmod),
+                    changefreq=changefreq,
+                    priority=priority,
                 )
-                self.write_url(fake, fd)
-
-            # add template pages
-            # We use items for Py3k compat. .iteritems() otherwise
-            for path, template_page_url in self.context["TEMPLATE_PAGES"].items():
+            )
 
-                # don't add duplicate entry for index page
-                if template_page_url == "index.html":
-                    continue
 
-                fake = FakePage(
-                    status="published",
-                    date=self.now,
-                    url=template_page_url,
-                    save_as=template_page_url,
-                )
-                self.write_url(fake, fd)
+def _check_config(settings):
+    config = settings.get("SITEMAP", {})
+    for key in config.keys():
+        if key not in ("format", "exclude", "priorities", "changefreqs"):
+            warning("sitemap: Invalid 'SITEMAP' key: {!r}".format(key))
+    for key in config.get("changefreqs", {}).keys():
+        if key not in CHANGEFREQ_DEFAULTS:
+            warning("sitemap: Invalid 'changefreqs' key: {!r}".format(key))
+    for key in config.get("priorities", {}).keys():
+        if key not in PRIORITY_DEFAULTS:
+            warning("sitemap: Invalid 'priorities' key: {!r}".format(key))
 
-            for page in pages:
-                self.write_url(page, fd)
 
-            if self.format == "xml":
-                fd.write(XML_FOOTER)
+def initialize(pelican):
+    _check_config(pelican.settings)
+    fmt = _get_format(pelican.settings)
+    filename = os.path.join(pelican.output_path, "sitemap." + fmt)
+    with open(filename, "w") as f:
+        f.truncate()
 
 
-def get_generators(generators):
-    return SitemapGenerator
+def finalize(pelican):
+    fmt = _get_format(pelican.settings)
+    filename = os.path.join(pelican.output_path, "sitemap." + fmt)
+    if fmt == "xml":
+        with open(filename, "r+") as f:
+            contents = f.read()
+            f.seek(0)
+            f.truncate()
+            f.write(XML_FILE_TEMPLATE.format(contents))
```

Comment:
This is strange. You write to file before, now read it, rewind and write a slightly different one.